### DISCUSSION TITLE
Hi, how can I model a elevator in the AGV control system based on OpenTCS ?

### DISCUSSION THREAD

**Discussion by hitmangithub** [openTCS/opentcs#113] [General]
Hi, how can I model a elevator in the AGV control system based on OpenTCS ?
Use a location with a multi-operation locationtype? Can someone give me some suggestioins or examples?
Thanks

**Comment by swltr** [comment]
@hitmangithub An elevator is one use case for peripheral devices. You'd have to implement a peripheral device driver and use peripheral operations at paths to call / wait for the elevator.

**Reply by vincentknw** [reply]
i found its difficult to model an elevator, if an elevator have 4 floor, there will be 6 paths to add, we have to assign a peripheral to each path,and when agv pass through the path ,it cannot move,but model coordinate of each layer may be diffrent。did i use it in a wrong way,can you offer an example,Finally, thank you again for your contribution!
![1713840769080](https://github.com/openTCS/opentcs/assets/107831846/0d319abf-0ff1-424b-832a-e1f52de32cef)

**Reply by swltr** [reply]
@vincentknw From my understanding of your description, it's more or less correct. You don't need a separate peripheral for every path, though. One elevator could be modeled as a single peripheral.

**Reply by vincentknw** [reply]
as for an elevator path, peripheral device(elevator) need to know the source layer and target layer, path configured with a peripheral operation,but cannot set the layer infomation,so i have to config each periheral device with layer properties,and the adapter can get layer infomation from properties.

**Reply by swltr** [reply]
@vincentknw Why not model the elevator as a single point, with four paths leading to this point and four paths leading out of it? You could then encode the information which floor the elevator is supposed to move to as part of the peripheral operation, e.g. `"move-to-layer:3"` to go to layer three.

**Reply by vincentknw** [reply]
> @vincentknw Why not model the elevator as a single point, with four paths leading to this point and four paths leading out of it? You could then encode the information which floor the elevator is supposed to move to as part of the peripheral operation, e.g. `"move-to-layer:3"` to go to layer three.

elevator coordinate may be different on different floor,because robot navigate with slam map , there are some cumulative error especially in large scenes with multiple elevators when mapping, so i have to model different point in different layer.But its a good idea to encode the information of peripheral operation, i will try it，thank you.

**Comment by winterdongsh** [comment]
Is my understanding correct below:
elevator modeling in OpenTCS
define path1’s peripheral action：
1 after_allocation, execute elevator-0001.to_level1, must be finished
2 after_movement, execute elevator-001.to_level2, must be finished
define path2’s peripheral action:
1 after_allocation, execute elevator-0001.to_level2, must be finished
2 after_movement, execute elevator-001.to_level1, must be finished
Note: If the elevator has a door, you can also add actions to open and close the elevator door.
<img width="248" alt="image" src="https://github.com/openTCS/opentcs/assets/29852730/7e5934c7-9982-4376-904b-7bbb3ba737ca">
